### ISSUE TITLE
Softmax interleaved attention optimized should use block_size=4 or 8 to maximize dest usage

### ISSUE THREAD

**Issue by rmillerTT** [tenstorrent/tt-metal#39050]
Currently, `SoftmaxShardedProgramFactoryAttentionOptimized` is computing `block_size` using the `find_max_divisor()` method. This underutilizes the 4 (if `fp32_acc_en`) or 8 (otherwise) dest registers in many cases. The kernels then can use `generic::blocks()` to iterate over the blocks with bounds checking.

Layernorm/RMSNorm was already adapted to this pattern, so it can be used as a prototype. 

Here's a Claude Code plan that can get you started:

 Ready to code?                                                                                                           
                                                                                                                          
 Here is Claude's plan:                                                                                                   
╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌
 Plan: Convert Softmax Attention Optimized to generic::blocks() with Fixed Block Size                                   
                                                                                                                        
 Context

 The softmax attention optimized factory currently computes block_size via find_max_divisor(Wt, 4/8), which finds the
 largest divisor of Wt up to 4 (fp32) or 8 (fp16). This guarantees Wt % block_size == 0 but results in unpredictable
 block sizes (e.g., Wt=195 -> block_size=3). The layernorm factory uses a simpler block_size = fp32_dest_acc_en ? 4 : 8
 with generic::blocks() to handle partial last blocks. This plan ports that pattern to softmax.

 Files to Modify

 1. Factory: ttnn/cpp/ttnn/operations/normalization/softmax/device/softmax_program_factory_attention_optimized.cpp
 2. Standard compute: .../kernels/attention/compute/softmax.cpp
 3. Large tensor compute: .../kernels/attention/compute/softmax_large_tensor.cpp
 4. Standard reader: .../kernels/attention/dataflow/reader_unary_interleaved_sm.cpp
 5. Large tensor reader: .../kernels/attention/dataflow/reader_unary_interleaved_sm_large_tensor.cpp
 6. Writer: .../kernels/attention/dataflow/writer_unary_interleaved_start_id_blocked_sm.cpp

 Reference: blocked_range.h at ttnn/cpp/ttnn/operations/normalization/kernel_util/generic/blocked_range.h

 ---
 1. Factory (softmax_program_factory_attention_optimized.cpp)

 create() changes

 - Line 73-74: block_size = fp32_dest_acc_en ? 4 : 8; (replace find_max_divisor)
 - CB sizing — replace div_up(Wt, block_size) * block_size with tt::round_up(Wt, block_size):
   - in0_t (line 77): numeric_stable ? tt::round_up(Wt, block_size) : block_size * 2
   - in4_t (line 83): tt::round_up(Wt, block_size)
   - im4_t (line 87): tt::round_up(Wt, block_size)
   - im0_t (line 90): tt::round_up(Wt, block_size)
   - im3_t (line 95): tt::round_up(Wt, block_size) + block_size
 - Remove/update assertions:
   - Line 91: im0_t == Wt -> remove (im0_t >= Wt now)
   - Line 122: im3_t == Wt + block_size -> im3_t == tt::round_up(Wt, block_size) + block_size
   - Line 124: Wt % block_size == 0 -> REMOVE
   - Line 125: block_size != -1 -> REMOVE
 - Writer runtime args (line 367): Change layout to pass Wt separately:
   - Old: {out_addr, num_tile_rows_per_core * Wt, tile_offset, block_size, mask_padded_data, num_datum_padded}
   - New: {out_addr, num_tile_rows_per_core, Wt, tile_offset, block_size, mask_padded_data}
 - Zero-core writer args (line 299): Match new layout: {0, 0, 0, 0, 0, 0}

 override_runtime_arguments() changes

 - Line 406-407: block_size = fp32_dest_acc_en ? 4 : 8;
 - CB sizing (lines 428-439): Same round_up changes
 - Remove assertions (lines 434, 439, 445-446)
 - Writer args (lines 576-581): Match new layout:
   - [1] = num_tile_rows_per_core, [2] = Wt, [3] = tile_offset, [4] = block_size, [5] = mask_padded_data
 - Zero-core writer (line 531): writer_kernel_args[1] = 0 (still zeroes num_tile_rows)

 ---
 2. Standard Compute Kernel (softmax.cpp)

 Add at top:
 #include "ttnn/operations/normalization/kernel_util/generic/blocked_range.h"
 namespace generic = norm::kernel_util::generic;

 Add after reading runtime args:
 const auto wt_blocks = generic::blocks(Wt, ndst);
 const auto total_wt = wt_blocks.total_with_remainder();

 Loop conversions (8 blocked loops, 2 cumulative loops)

 Cumulative reduction loops — KEEP AS-IS:
 - calc_numeric_stable line 31: for (wt = 0; wt < Wt; wt++) (MAX reduce)
 - kernel_main line 277: for (wt = 0; wt < Wt; wt++) (SUM reduce)

 Blocked loops — convert to generic::blocks():

 Each for (wt = 0; wt < Wt; wt += ndst) { ... for (wt8 = 0; wt8 < ndst; wt8++) } becomes for (auto block : wt_blocks) {
 ... for (auto wt8 : block.local()) } with:
 - CB reserve/push/pop: ndst -> block.full_block_size()
 - Inner tile count: ndst -> block.local() (handles partial last block)
 - Global index: wt + wt8 -> block.to_global(wt8)

 Specific loops:
 1. calc_numeric_stable lines 49-65 (exp(x-max)): Convert. Change cb_pop_front(cb_in, Wt) to total_wt. Change
 cb_wait_front(cb_out, Wt) to total_wt.
 2. Lines 115-131 (fused scale): Convert.
 3. Lines 143-174 (attention mask + exp): Convert. Cumulative mask wait wt + ndst -> block.start() +
 block.full_block_size().
 4. Lines 205-233 (mask_padded_data path): Convert. Last-tile detection wt == (Wt - ndst) && (wt8 == ndst - 1) ->
 block.to_global(wt8) == Wt - 1.
 5. Lines 247-266 (no-mask, no-numeric-stable exp): Convert.
 6. Lines 303-318 (final multiply): Convert. Uses global index into cumulative cb_exps.

 Pop-all changes:
 - cb_pop_front(cb_exps, Wt) (line 320) -> total_wt
 - cb_pop_front(cb_fused_attn, Wt) (lines 183, 190) -> total_wt
 - cb_pop_front(cb_in, Wt) in calc_numeric_stable (line 66) -> pass total_wt to the function or compute inside

 ---
 3. Large Tensor Compute (softmax_large_tensor.cpp)

 Add same include + namespace alias.

 Convert 5 helper functions that use for (cur_blk = 0; cur_blk < cb_length_t; cur_blk += blk) with manual remainder (if
 (cb_length_t - cur_blk < blk) { blk = ... }) to for (auto block : generic::blocks(cb_length_t, blk)):

 1. apply_fused_scale_mask (lines 80-110)
 2. apply_fused_attn_mask (lines 111-158) — last-tile mask check: cur_blk == cb_length_t - blk -> check via
 block.to_global(cur_dst) == cb_length_t - 1
 3. pad_input (lines 161-191) — same last-tile adjustment
 4. exp_cb (lines 193-239) — remove ASSERT(cb_length_t % blk == 0) at line 199
 5. apply_recip (lines 290-314)

 reduce_cb (per-tile) — KEEP AS-IS.

 ---
 4. Standard Reader (reader_unary_interleaved_sm.cpp)

 Add include + namespace alias.

 Convert 3 loops that iterate for (j = 0; j < Wt; j += blk) with inner for (r = 0; r < blk; r++):

 1. Lines 66-78: Input tile read -> for (auto block : generic::blocks(Wt, blk))
 2. Lines 85-95: Causal mask read -> same
 3. Lines 107-119: Non-causal mask read -> same

 Pattern per loop:
 for (auto block : generic::blocks(Wt, blk)) {
     cb_reserve_back(cb, block.full_block_size());
     uint32_t l1_write_addr = get_write_ptr(cb);
     for (auto r : block.local()) {
         noc_async_read_tile(curr_tile++, accessor, l1_write_addr);
         l1_write_addr += tile_bytes;
     }
     noc_async_read_barrier();
     cb_push_back(cb, block.full_block_size());
 }

 This is a correctness fix: without blocks(), reading past the tensor boundary when Wt % blk != 0.

 ---
 5. Large Tensor Reader (reader_unary_interleaved_sm_large_tensor.cpp)

 Add include + namespace alias.

 Convert lines 80-103: Same pattern as standard reader. The read loop per pass becomes for (auto block :
 generic::blocks(Wt, blk)).

 ---
 6. Writer (writer_unary_interleaved_start_id_blocked_sm.cpp)

 Add include + namespace alias.

 Change runtime arg layout:
 const uint32_t dst_addr = get_arg_val<uint32_t>(0);
 const uint32_t num_tile_rows = get_arg_val<uint32_t>(1);  // was num_tiles
 const uint32_t Wt = get_arg_val<uint32_t>(2);             // NEW
 const uint32_t tile_offset = get_arg_val<uint32_t>(3);     // was [2]
 const uint32_t blk = get_arg_val<uint32_t>(4);             // was [3]
 const uint32_t mask_padded_data = get_arg_val<uint32_t>(5); // was [4]

 Convert write loop (lines 47-58) from flat to row-by-row:
 uint32_t tile_id = tile_offset;
 for (uint32_t row = 0; row < num_tile_rows; row++) {
     for (auto block : generic::blocks(Wt, blk)) {
         cb_wait_front(cb_id_out0, block.full_block_size());
         uint32_t l1_read_addr = get_read_ptr(cb_id_out0);
         for (auto j : block.local()) {
             noc_async_write_tile(tile_id++, s, l1_read_addr);
             l1_read_addr += tile_bytes;
         }
         noc_async_write_barrier();
         cb_pop_front(cb_id_out0, block.full_block_size());
     }
 }

 ---
 Verification

 1. Build: ./build_metal.sh (kernel source changes are JIT-compiled, but factory C++ is in _ttnncpp.so)
 2. Clear JIT cache: rm -rf ~/.cache/tt-metal-cache/
 3. Run existing softmax tests: python -m pytest tests/ttnn/unit_tests/operations/test_softmax.py -xvs
 4. Test the specific case from the conversation: Shape [1,1,32,6240], BF16, fp32_dest_acc_en=True, numeric_stable=True
 (Wt=195, not divisible by 4 — exercises partial block handling)
 5. Test with small Wt not divisible by 4 or 8 (e.g., Wt=3, Wt=5, Wt=7)
 6. Test with FUSED_SCALE_MASK and CAUSAL_MASK variants if covered by existing tests
